### ISSUE TITLE
Convert to a Lerna monorepo

### ISSUE THREAD

**Issue by Lange** [nodecg/nodecg#299]
https://lernajs.io/

[Lerna](https://lernajs.io/) is a command-line tool for managing a monorepo. A monorepo is a git repository that contains multiple npm packages. The main selling point of Lerna is that it lets you split your code up into separate easy-to-test modules without having to put each of them into their own GitHub repo, which would increase maintenance overhead.

Lerna could be applied to NodeCG to move the following repos back into this main repo:
- https://github.com/nodecg/logger
- https://github.com/nodecg/bundle-manager
- https://github.com/nodecg/nodecg-bundle-parser

I think that this could make maintaining these 3 modules easier, and reduce the effort required to break even more of NodeCG's functionality out into small, easy-to-test modules.

@MattMcNam I'd like to get your thoughts on this.

Input from others is welcome also!

**Comment by MattMcNam** [comment]
👍 👍 👍 

I'll have a quick read through Lerna now, but in general I'd like to go to a monorepo. Switching between those 3 repos and fiddling with npm link is frustrating.

**Comment by Lange** [comment]
I'm messing with using Lerna now for that OBS widget that I've been talking about on Twitter. It's definitely the right move and it helps in a lot of ways, but Lerna does make some assumptions about how you do things like testing and code coverage that can make the initial setup a bit complicated. Not impossible, just complicated.

That initial set up is just one-time work though, and it seems worth it so far.

**Comment by Lange** [comment]
Now that I have experience using a Lerna monorepo in production thanks to https://github.com/nodecg/nodecg-obs, I'm going to say that we commit to this and do this for v0.9.

**Comment by Lange** [comment]
I started work on this and immediately ran into something that made me pause.

When I work on NodeCG bundles, it's already kind of weird that I'm working on them within a subdirectory of another repository (`nodecg/bundles/`). Converting NodeCG to a monorepo makes that even stranger, and makes NodeCG installations even weirder. Trying to explain to someone how to install NodeCG and install bundles in this new format would be weird. And in my use case, all my bundles would be in `nodecg/packages/nodecg/bundles` (or `nodecg/packages/nodecg-core/bundles`).

This isn't a showstopping issue, but it needs more thought before I dig in and start doing this work.

**Comment by Lange** [comment]
This is maybe a good time to talk about what an ideal NodeCG workflow looks like. I don't yet have a clear grasp of what that is, but here's what I think some of the attributes would be:
- I want to be able to easily choose what version of NodeCG to run, even down to a specific commit hash
- It'd be cool to not have to worry about remembering to run `npm install` and `bower install` on NodeCG itself
- It might be nice if `nodecg` was just a command line program, and that a bundle or set of bundles could be spun up in a NodeCG instance from anywhere on the filesystem.
  - Example A: I have a `Documents/my-bundle` folder. If I `cd` into that and run `nodecg`, NodeCG should start up and load that bundle
  - Example B: I have a `Documents/bundles` folder that contains multiple bundles. If I run `nodecg` from this folder, NodeCG would load all bundles in that folder (obeying the whitelist/blacklist from the config)
  - I don't know where the NodeCG configuration and the per-bundle configuration would go in this type of setup.

In short, I think that turning NodeCG into a monorepo starts to really expose the fact that the current way that NodeCG and bundles are installed is weird, and that there is room for a fundamental shift in how we approach that.

**Comment by Lange** [comment]
Revisiting this. I'd like to put more thought to this and get input from others, particularly in regard to my previous comment.

**Comment by Lange** [comment]
It's possible that I'm barking up the wrong tree entirely here. Maybe a monorepo isn't the answer. The only value that the NodeCG project gains from splitting its code up into multiple packages is that it becomes easier to test them. However, this quality could also be gained by just structuring the code well in a single package.

In the year or two since these other 3 modules have been broken out into separate packages, I don't think its ever once come in handy. It's only ever been a hindrance that adds additional overhead.

I'll have to think more about what re-structuring NodeCG to be one package again (while still maintaining the modular, easy-to-test nature of the other packages) looks like.

**Comment by Lange** [comment]
Alright, so I think I've decided that a monorepo is not the way to go, and that instead I should be focusing on just merging everything back into one package, while maintaining good coding practices that let the modules remain isolated and easily testable.

However, when I started trying to do this, I ran into issues when running tests. A lot of the existing tests for these other 3 packages are incompatible with NodeCG's tests, due to the fact that Mocha runs everything in one shared process with a shared global state.

To address this, I have completely re-written NodeCG's tests to use `ava` in #322. We lost a few small tests in the process, but overall I think its worth it and those tests can always be re-written later. I think I now have a path forward for merging these 3 packages back into NodeCG core.

**Comment by Lange** [comment]
Closed by #323.
